lib/ecto/adapters/mssql/connection.ex

```diff
@@ -0,0 +1,1644 @@
+if Code.ensure_loaded?(Tds) do
+  defmodule Ecto.Adapters.MsSql.Connection do
+    @moduledoc false
+    require Logger
+    alias Tds.Query
+    alias Ecto.Query.Tagged
+    require Ecto.Schema
+
+    @behaviour Ecto.Adapters.SQL.Connection
+    # @behaviour Ecto.Adapters.SQL.Query
+
+    @unsafe_query_strings ["'\\"]
+
+    @impl true
+    @doc false
+    def child_spec(opts) do
+      opts
+      |> Keyword.put_new(:use_elixir_calendar_types, true)
+      |> Tds.child_spec()
+    end
+
+    @impl true
+    def prepare_execute(pid, _name, statement, params, opts \\ []) do
+      query = %Query{statement: statement}
+
+      {params, _} =
+        Enum.map_reduce(params, 1, fn param, acc ->
+          {value, type} = prepare_param(param)
+          {%Tds.Parameter{name: "@#{acc}", value: value, type: type}, acc + 1}
+        end)
+
+      opts = Keyword.put(opts, :parameters, params)
+      DBConnection.prepare_execute(pid, query, params, opts)
+    end
+
+    @impl true
+    def execute(pid, statement, params, opts) when is_binary(statement) or is_list(statement) do
+      query = %Query{statement: statement}
+
+      {params, _} =
+        Enum.map_reduce(params, 1, fn param, acc ->
+          {value, type} = prepare_param(param)
+          {%Tds.Parameter{name: "@#{acc}", value: value, type: type}, acc + 1}
+        end)
+
+      opts = Keyword.put(opts, :parameters, params)
+
+      case DBConnection.prepare_execute(pid, query, params, opts) do
```

Comment:
Do you need to pass `params` both as the third argument here AND inside `opts`? The same is done in the function below.